pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,552 @@
+	cachedciImps := c.installedCIImports.List()
+	for _, ciImp := range cachedciImps {
+		cached := ciImp.(*ciImportInfo)
+		key := getCacheKeyWithPrefix(gwName + cached.ciImport.Spec.ClusterID)
+		if err := c.ofClient.UninstallMulticlusterFlows(key); err != nil {
+			return fmt.Errorf("failed to uninstall multi-cluster flows to remote Gateway in ClusterInfoImport %s: %v",
+				cached.ciImport.Name, err)
+		}
+		c.installedCIImports.Delete(ciImp)
+	}
+	return nil
+}
+
+func (c *MCRouteController) getAllCIImports(ciImp *mcv1alpha1.ClusterInfoImport) ([]*mcv1alpha1.ClusterInfoImport, error) {
+	var allciImports []*mcv1alpha1.ClusterInfoImport
+	var err error
+	if ciImp == nil {
+		allciImports, err = c.ciImportLister.ClusterInfoImports(c.namespace).List(labels.Everything())
+		if err != nil {
+			return nil, fmt.Errorf("failed to get ClusterInfoImport list, err: %v", err)
+		}
+		if len(allciImports) == 0 {
+			klog.InfoS("No valid remote ClusterInfo imported, do nothing")
+			return nil, nil
+		}
+	} else {
+		allciImports = append(allciImports, ciImp)
+	}
+	return allciImports, nil
+}
+
+func (c *MCRouteController) getGateways() ([]*mcv1alpha1.Gateway, error) {
+	gws, err := c.gwLister.Gateways(c.namespace).List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Failed to get Gateway list", "namespace", c.namespace)
+		return nil, err
+	}
+	if len(gws) == 0 {
+		return nil, nil
+	}
+	return gws, nil
+}
+
+// getActiveGateway compares Gateway's CreationTimestamp to get the active Gateway,
+// The last created Gateway will be the active Gateway.
+func (c *MCRouteController) getActiveGateway(gws []*mcv1alpha1.Gateway) *mcv1alpha1.Gateway {
+	if len(gws) == 0 {
+		return nil
+	}
+	// Comparing Gateway's CreationTimestamp to get the last created Gateway.
+	lastCreatedGW := gws[0]
+	for _, gw := range gws {
+		if lastCreatedGW.CreationTimestamp.Before(&gw.CreationTimestamp) {
+			lastCreatedGW = gw
+		}
+	}
+	return lastCreatedGW
+}
+
+// iAmGateway checks if the current Node is a Gateway.
+func (c *MCRouteController) iAmGateway(gws []*mcv1alpha1.Gateway) bool {
+	var iAmGW bool
+	for _, gw := range gws {
+		if gw.Name == c.nodeConfig.Name {
+			iAmGW = true
+			break
+		}
+	}
+	return iAmGW
+}
+
+func generatePeerConfigs(subnets []string, gatewayIP net.IP) (map[*net.IPNet]net.IP, error) {
+	peerConfigs := make(map[*net.IPNet]net.IP, len(subnets))
+	for _, subnet := range subnets {
+		_, peerCIDR, err := net.ParseCIDR(subnet)
+		if err != nil {
+			klog.Errorf("Failed to parse subnet %s", subnet)
+			return nil, err
+		}
+		peerConfigs[peerCIDR] = gatewayIP
+	}
+	return peerConfigs, nil
+}
+
+// getPeerGatewayIP will always return the first Gateway IP.
+func getPeerGatewayIP(spec mcv1alpha1.ClusterInfo) net.IP {
+	if len(spec.GatewayInfos) == 0 {
+		return nil
+	}
+	return net.ParseIP(spec.GatewayInfos[0].GatewayIP)
+}
+
+func getCacheKeyWithPrefix(key string) string {
+	return "mc-" + key
+}
```

Comment:
If I understand correctly, the whole controller's logic can be simplified to:
```
func syncFlows() error {
	activeGW := ...

	// Active gateway is never set.
	if activeGW == nil && c.installedActiveGWName == "" {
		return nil
	}
	
	// Active gateway doesn't change.
	if activeGW != nil && activeGW.Name == c.installedActiveGWName {
		// Sync inter-cluster flows if ClusterInfoImports change.
		if err := c.syncInterClusterFlows(); err != nil {
			return err
		}
		return nil
	}

	if c.installedActiveGWName != "" {
		if err := c.cleanupFlows(c.installedActiveGWName); err != nil {
			return err
		}
	}

	if activeGW != nil {
		if activeGW.Name == c.nodeConfig.Name {
			err = c.installFlowsOnGateway(activeGW)
		} else {
			err = c.installFlowsOnRegularNode(activeGW)
		}
		if err != nil {
			return err
		}
		c.installedActiveGWName = activeGW.Name
	} else {
		c.installedActiveGWName = ""
	}

	return nil
}

func syncInterClusterFlows() error {
	installedCIImports := c.installedCIImports.List()
	desiredCIImports := c.ciImportLister.ClusterInfoImports(c.namespace).List(labels.Everything())
	
	for ciImport := range desiredCIImports {
		if installedCIImports !has ciImport {
			addMCNodeFlows(ciImport)
		}
		c.installedCIImports.Add(ciImport)
	}
	
	for ciImport := range installedCIImports {
		if desiredCIImports !has ciImport {
			deleteMCNodeFlows(ciImport)
		}
		c.installedCIImports.Delete(ciImport)
	}
	return nil
}
```

All event handlers (both gateway and ClusterInforImports) should just enqueue a same dummy key to the queue to trigger a full-resync of the flows.

Reply:
This makes sense! The only thing is we need to think about how to handle partial failure of flow installation/cleanup.

Reply:
Make sense! Again, the only thing is we need to handle partial success of flow installation.

Reply:
Thanks, I have refined the codes with this new structure. Could you please review again? thanks.